cache.go

```diff
@@ -0,0 +1,839 @@
+	var e error
+	s3Clnt := &S3CacheClient{}
+	options := minio.Options{
+		Creds:        creds,
+		Secure:       cfg.useTLS,
+		Region:       "",
+		BucketLookup: 0,
+	}
+	u, err := url.Parse(cfg.endpoint)
+	if err != nil {
+		return nil
+	}
+	api, e := minio.NewWithOptions(u.Host, &options)
+	if e != nil {
+		return nil
+	}
+	tr := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		MaxIdleConns:          1024,
+		MaxIdleConnsPerHost:   1024,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		// Set this value so that the underlying transport round-tripper
+		// doesn't try to auto decode the body of objects with
+		// content-encoding set to `gzip`.
+		//
+		// Refer:
+		//    https://golang.org/src/net/http/transport.go?h=roundTrip#L1843
+		DisableCompression: true,
+	}
+
+	if cfg.useTLS {
+		// Keep TLS config.
+		tlsConfig := &tls.Config{
+			RootCAs: mustGetSystemCertPool(),
+			// Can't use SSLv3 because of POODLE and BEAST
+			// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+			// Can't use TLSv1.1 because of RC4 cipher usage
+			MinVersion:         tls.VersionTLS12,
+			NextProtos:         []string{"http/1.1"},
+			InsecureSkipVerify: ctx.GlobalBool("insecure"),
+		}
+		tr.TLSClientConfig = tlsConfig
+	}
+
+	var transport http.RoundTripper = tr
+	// Set the new transport.
+	api.SetCustomTransport(transport)
+	// Set app info.
+	api.SetAppInfo(ctx.App.Name, ctx.App.Version)
+	// Store the new api object.
+	s3Clnt.api = api
+	cfg.endpoint = strings.TrimSuffix(cfg.endpoint, slashSeparator)
+
+	target, err := url.Parse(cfg.endpoint)
+	if err != nil {
+		console.Fatalln(fmt.Errorf("Unable to parse input arg %s: %s", cfg.endpoint, err))
+	}
+	if target.Scheme == "" {
+		target.Scheme = "http"
+	}
+	if target.Scheme != "http" && target.Scheme != "https" {
+		console.Fatalln("Unexpected scheme %s, should be http or https, please use '%s --help'",
+			cfg.endpoint, ctx.App.Name)
+	}
+	if target.Host == "" {
+		console.Fatalln(fmt.Errorf("Missing host address %s, please use '%s --help'",
+			cfg.endpoint, ctx.App.Name))
+	}
+	s3Clnt.methods = []string{http.MethodGet, http.MethodHead}
+	s3Clnt.bucket = cfg.bucket
+	s3Clnt.minSize = cfg.minSize
+	s3Clnt.healthCheckDuration = cfg.duration
+	s3Clnt.endpoint = cfg.endpoint
+	s3Clnt.useTLS = target.Scheme == "https"
+	s3Clnt.httpClient = &http.Client{Transport: tr}
+	go s3Clnt.healthCheck()
+	return s3Clnt
+}
+func sortURLParams(URL *url.URL) {
+	params := URL.Query()
+	for _, param := range params {
+		sort.Slice(param, func(i, j int) bool {
+			return param[i] < param[j]
+		})
+	}
+	URL.RawQuery = params.Encode()
+}
+
+func generateKey(URL *url.URL, host string) string {
+	keyName := fmt.Sprintf("%s%s", host, URL.String())
```

Comment:
this function might have heavy allocations `generateKey()` due to `fmt.Sprintf` usage can you add a `bench` test for this?

Reply:
ah rather than doing `fmt.Sprintf` you can simply do like this 

```
func generateKey(u *url.URL, host string) string {
       hash := sha256.New()
       hash.Write([]byte(host))
       hash.Write([]byte(u.String())
       hashSum := hex.EncodeToString(hash.Sum(nil))
       return hashSum[0:2] + slashSeparator + hashSum[2:4] + slashSeparator + hashSum
}
```

Reply:
thanks,done